k2/csrc/array_ops_inl.h

```diff
@@ -877,6 +882,32 @@ void Assign(Array2<T> &src, Array2<T> *dest) {
   }
 }
 
+template <typename T>
+Array1<T> MergeWithMap(const Array1<uint32_t> &merge_map,
+                       int32_t num_srcs, const Array1<T> **src) {
+  int32_t dim = merge_map.Dim();
+  ContextPtr &c = merge_map.Context();
+  std::vector<const T*> src_ptrs_vec(num_srcs);
+  int32_t src_tot_dim = 0;
+  for (int32_t i = 0; i < num_srcs; i++) {
+    src_tot_dim += src[i]->Dim();
+    src_ptrs_vec[i] = src[i]->Data();
+  }
+  Array1<const T*> src_ptrs(c, src_ptrs_vec);
```

Comment:
I believe you want to check src_tot_dim == dim here, as you have computed src_tot_dim above but did not use it.